ext/js/comm/anki-connect.js

```diff
@@ -579,6 +590,43 @@ export class AnkiConnect {
         return /** @type {T[]} */ (result);
     }
 
+    /**
+     * @param {unknown} result
+     * @param {number} expectedCount
+     * @param {string} [context]
+     * @returns {({ canAdd: true } | { canAdd: false, error: string })[]}
+     * @throws {Error}
+     */
+    _normalizeCanAddNotesWithErrorDetailArray(result, expectedCount, context) {
+        if (!Array.isArray(result)) {
+            throw this._createUnexpectedResultError('array', result, '');
+        }
+        if (expectedCount < 0) {
+            expectedCount = result.length;
+        } else if (expectedCount !== result.length) {
+            throw this._createError(`Unexpected result array size${context}: expected ${expectedCount}, received ${result.length}`, result);
+        }
+        for (let i = 0; i < expectedCount; ++i) {
+            const item = /** @type {unknown} */ (result[i]);
+            if (item === null || typeof item !== 'object') {
+                throw this._createError(`Unexpected result type at index ${i}: expected object, received ${this._getTypeName(item)}`, result);
+            }
+
+            const {canAdd, error} = /** @type {{[key: string]: unknown}} */ (item);
+            if (typeof canAdd !== 'boolean') {
+                throw this._createError(`Unexpected result type at index ${i}, field canAdd: expected boolean, received ${this._getTypeName(canAdd)}`, result);
+            }
+
+            if (canAdd && typeof error !== 'undefined') {
+                throw this._createError(`Unexpected result type at index ${i}, field error: expected undefined, received ${this._getTypeName(error)}`, result);
+            }
+            if (!canAdd && typeof error !== 'string') {
+                throw this._createError(`Unexpected result type at index ${i}, field error: expected string, received ${this._getTypeName(error)}`, result);
+            }
+        }
+        return /** @type {({ canAdd: true } | { canAdd: false, error: string })[]} */ (result);
```

Comment:
Preferably don't return `result` directly. Rather, construct a new array after validating each item, and manually add each entry as a new object. This is safer, since it makes sure that any extra data passed from the AnkiConnect API is _never_ passed along into other places in the app. E.g. if AnkiConnect for whatever reason returned `{canAdd: false, error: 'something', otherData: <someNonsense>}`, all other fields will never be exposed. This should require less type casting as well.

Reply:
done